source/arm9/libc/fatfs/cache.c

```diff
@@ -22,33 +23,48 @@ static cache_entry_t *cache_entries;
 static uint8_t *cache_mem;
 static uint32_t cache_num_sectors;
 
-int cache_init(uint32_t num_sectors)
+extern uint8_t* dldiGetStubDataEnd(void);
+extern uint8_t* dldiGetStubEnd(void);
+
+int cache_init(int32_t num_sectors)
 {
     // If this function is called after the first time, clear the cache and
     // allocate a new one.
 
     if (cache_entries != NULL)
         free(cache_entries);
 
-    if (cache_mem != NULL)
+    if (cache_mem != NULL && (uintptr_t)cache_mem > (uintptr_t)dldiGetStubEnd())
         free(cache_mem);
 
-    cache_entries = calloc(num_sectors, sizeof(cache_entry_t));
-    if (cache_entries == NULL)
-        return -1;
+    int32_t stub_space_sectors = (dldiGetStubEnd() - dldiGetStubDataEnd()) >> 9;
+    if (num_sectors < 0) {
+        num_sectors = stub_space_sectors;
```

Comment:
Hmmm is this a good idea? You're allocating less space for the cache of bigger DLDI drivers.

I think we should do this in a different way. Instead of having a big RAM chunk that we use as cache, we can have an array of pointers to sectors. Then, we allocate as many sectors as we can in the space of the DLDI driver, and the rest in the heap.

Reply:
Well, we don't need an array of pointers, maybe just two pointers: One to the free space in the DLDI driver and another one to the extra space in the heap.

Reply:
We can get away with just two pointers, yes.

>Hmmm is this a good idea? You're allocating less space for the cache of bigger DLDI drivers.

This is deliberate.

a) The vast majority of currently used DLDI drivers are <=8KB, which leaves a cache of 16-24 sectors.
b) On DSi, the DLDI driver is populated by a stub, which leaves a cache of 31 sectors.

The previous default cache size was 40 sectors; as the cache has been optimized to only be invoked for FAT directory/cluster table reads, I don't think it needs to be as large to get the same performance. Unless there's a significant performance gain from a larger cache, I think it is fine to make this approach the default (as using the DLDI reserved space for a sector cache is always "free", RAM-wise). Nevertheless, allocating larger caches is always possible - be it by defining your own `__dldi_size` symbol to reserve a 32KB area, or by passing a larger size to fatInit().